jlm/rvsdg/region.hpp

```diff
@@ -763,6 +773,34 @@ class Region
   region_bottom_node_list BottomNodes_;
   region_top_node_list TopNodes_;
   region_nodes_list Nodes_;
+  RegionObserver * observers_ = nullptr;
+
+  friend class RegionObserver;
+  friend class SimpleNode;
+  friend class StructuralNode;
+  friend class Input;
+};
+
+class RegionObserver
+{
+public:
+  ~RegionObserver();
+  RegionObserver(Region * region);
```

Comment:
Since we use pointers to these observers and their fields, I think it makes sense to give them deleted copy constructors and move constructors, or alternatively implement a move constructor that updates the pointers

Reply:
The region itself can update the head observer's pprev pointer in its move constructor as well

Reply:
I rather made it non-copyable (and non-movable) -- this should really semantically be a "pinned" class.